### PR TITLE
spike(pronunciation): settle on-device stress detection (#2359)

### PR DESCRIPTION
**Pull request by OPS-PIvers**
Resolves [#2359 — Detecting stress without the phoneme model](https://github.com/OPS-PIvers/SpartBoard/issues/2359), a decision ticket on the [Wayfinder Map: Multilingual Pronunciation Engine](https://github.com/OPS-PIvers/SpartBoard/issues/2331).

Opened by A1 on [#2335](https://github.com/OPS-PIvers/SpartBoard/issues/2335), which put lexical stress in scope and established that no CTC phoneme model can emit it.

**No feature code**, per the map's plan-don't-do rule. Decisions S1–S5 live in `scripts/spikes/stress-detection/DECISIONS.md`, pinned as **26 tests that run in CI** alongside the alignment harness (53 total across both spikes).

## Two premises checked against the artifact, both wrong

Each took under a minute to settle by reading the thing rather than reasoning about it — the map's rule 6.

**Frame timings exist, but not on the documented path.** `pipeline('automatic-speech-recognition')` routes wav2vec2 to `_call_wav2vec2`, which returns `{ text }` and discards all timing — it's marked `// TODO use kwargs` and warns options are ignored, and `return_timestamps` is read only by the Whisper branch. Taking the happy path forecloses stress detection *silently*. But `Wav2Vec2ForCTC` is publicly exported and returns raw `logits[1, frames, 392]`, so owning the decode yields a frame span per symbol at ~20 ms.

→ **No second model download and no forced aligner.** The map's "a second artifact may now join it" note can be struck.

**The nucleus set is derived, not written from memory.** `deriveNuclei.mjs` generates `nuclei.json` from the model's committed `vocab.json`: **244 of 387 sound tokens** carry a syllable (239 vowel-headed + 3 syllabic consonants + 2 glide-initial). Three traps a hand-written rule falls into:

- Diphthongs are **single tokens** (`aɪ`, `aʊ`, `eɪ`, `ɔɪ`, `əʊ`), so English *time* is one syllable with no merging rule — but `n̩`, `l̩`, `r̩` are **vowel-less nuclei**, and "count the vowels" undercounts every German reducing `-en` ending (*gehen*, *Wagen*, *haben*).
- Only two consonant-initial tokens contain a vowel: **`ja` and `ju`** — both ordinary German words, so a head-character rule loses them frequently and silently.
- **A textbook IPA vowel inventory is not this vocabulary's.** `ɚ` (r-coloured schwa — *better*, *water*) and `ᵻ` (espeak's reduced high vowel — *roses*, *wanted*) sit outside it, plus `ä` and `ũ`. Omitting `ɚ` and `ᵻ` costs those words a syllable, **which shifts every stress index after it**. Found in review by auditing the heads of the tokens the rule *rejected* — re-reading the list could not have found it, because the list was what was wrong.

## Decisions

| | |
| --- | --- |
| **S1** | Stress is **one syllable index** ("the 2nd of 3"), not a per-syllable vector. Detection becomes a purely relative ranking within one word, so nothing needs the calibration the map defers. Rejected primary/secondary/none: the engine compares for equality, so if detection can't separate secondary from unstressed, every multi-syllable German and English word mismatches. |
| **S2** | Syllables come from the phoneme stream — nucleus tokens plus their CTC frame spans. No second stage, no download. |
| **S3** | Prominence is **mapped onto the target syllable through the alignment** the engine already computes. Counts are never compared. **S3a**: a prominent nucleus that is an insertion is absent evidence. **S3b**: so is one substituted over a target consonant. |
| **S4** | Four cues, all relative within the word: duration, loudness, pitch, and the reduced vowel already in the phoneme stream (English only — Spanish doesn't reduce). |
| **S5** | Uncertainty is reported by **saying nothing** — the detector returns null and the caller omits the stress field, reaching A9's existing degradation path. |

## What this changes in the #2335 contract

**A10b is superseded.** S3 removes the syllable-count comparison entirely, so the risk A10b explicitly priced — noisy segmentation manufacturing errors the student didn't make — is now structurally absent rather than bounded by a low weight. It also stops the stress dimension double-charging an extra syllable that A5/A6 already dock as insertions.

**A10 is narrowed but survives.** Full credit for any accepted regional variant is intact; `accepted` becomes `number[]` rather than `number[][]`, and stress scores 100 or 0.

**A9 and A10a are untouched.** The combined-score formula and both degradation paths are unchanged; S5 decides when to invoke them.

The #2335 spike is **annotated, not rewritten** — it's the accepted record of that ticket and its 27 tests are the pinned form of A1–A13, so A10 and A10b gain pointers to the new file and CI stays green. Feature code takes the new shape.

## Notes for review

- **Licensing landmine, same family as the espeak GPL-3 read.** The obvious pitch library, `pitchfinder` (229 KB), is **GPL-3**, as is `node-pitchfinder`. Either would reintroduce exactly the client-bundle problem D1 removed by moving G2P server-side. **`pitchy` is MIT at 33 KB.**
- **Nothing here is measured on real audio**, per rule 4. Cue weights ship tunable and validating them against the human clips in `pronunciation-bias-probe/` is the natural next spike. Pitch declination — the first syllable of an isolated word reading as high regardless — is a specific open question.
- **S3a and S3b were decided without asking**, and are flagged as such in the decision record rather than presented as settled.

## Verification

`pnpm run type-check`, `pnpm run lint` (`--max-warnings 0`) and `pnpm run format:check` all pass. Spike suites: **53 tests passing**. The nucleus set was mutation-checked — reverting to the naive head-character rule fails the German assertions.

### CODE REVIEW

**Comment by OPS-PIvers** [comment]
All four notes confirmed against the code and fixed in d2f4abc. One of them was hiding something considerably bigger.

## The `ɪ̈` note found a real bug in the derivation

The suggested fix — substitute plain `ɪ` — would have been wrong: `ɪ` is a full KIT vowel that carries stress in *bit*, so marking it reduced would rule out stress on **every** syllable containing it. So I dropped `ɪ̈` and instead pinned the *whole* reduced-vowel set against `vocab.json`, on the reasoning that a symbol-level typo deserves a symbol-level guard.

That guard immediately failed on a symbol I had just added — and the failure was in the nucleus derivation, not the test.

**The textbook IPA vowel inventory the derivation was built from is not this vocabulary's vowel inventory.** Four vowels sit outside it, and two are load-bearing for English:

| Symbol | | Cost of omitting it |
| --- | --- | --- |
| **`ɚ`** | r-coloured schwa | *better*, *water*, *teacher* each lose a syllable |
| **`ᵻ`** | espeak's reduced high vowel | *roses*, *wanted* each lose a syllable |
| `ä` | centralized [a] | — |
| `ũ` | nasalized [u] | — |

**A lost syllable shifts every stress index after it**, so this would have produced confidently wrong answers on ordinary English words rather than failing visibly.

I found it by auditing the heads of the tokens the rule **rejected**, rather than by re-reading the vowel list — the list is what was wrong, so re-reading it would never have surfaced this. That audit is the check worth repeating if the label set ever changes. `ʲ` (palatalization modifier) and `ɫ` (dark l) appear in the same listing and are correctly excluded; the syllabic form of l is the separate token `l̩`.

Nucleus set **239 → 244** (vowel-headed 234 → 239). `ᵻ` also joins the reduced-vowel set, being precisely the reduction the English cue exists to detect. Both counts and both worked examples are pinned as tests.

## The other three

- **`Nucleus.syllableIndex`** — removed. Your reading of the risk was right: it held a *spoken* nucleus index where a caller reaching for it would want the *target* syllable index `mapToTargetSyllable` returns. `phonemeIndex` is now the only index the type carries, with the reason recorded on the field, so there is no spoken-space number lying around to be mistaken for a target-space one.
- **Side-effectful `.find()`** — replaced with the explicit loop.
- **Untested `measurements.length !== nuclei.length` guard** — correct, the "no nuclei" case fired the earlier `nuclei.length === 0` guard and never reached it. Test added.

25 stress tests, **52 across both spikes**. Type-check and formatting clean; the spike directories are eslint-ignored, matching `alignment-engine/`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by OPS-PIvers** [comment]
Both fixed in 8fc4f39.

**Stale count** — correct, and mine: the JSDoc still said 239 after the set grew to 244 in the previous commit. The test and `DECISIONS.md` were already right, which is the worst version of this since the comment is exactly what a reader would cross-check the set against.

**Substituted nucleus over a target consonant** — verified: `countSyllables(target.slice(0, position))` is 0 for a nucleus aligned to a consonant, so it already returned `null` and degraded. Undocumented, though, and you were right that a future reader would have to work out whether that was deliberate.

Recorded as **S3b**, with a test. The reasoning I've written down: a student saying `a` where the target has `p` produced a vowel in the **onset** — it is not the first target syllable's nucleus, which is the `e` after it — so mapping the prominence to syllable 1 would assert a stress placement the student never made. The test also pins that nuclei later in the word still map normally; only the one sitting ahead of every target nucleus is unmappable.

That makes two edge cases now decided without being asked about (S3a, S3b), both flagged as such in the record rather than presented as settled.

26 stress tests, **53 across both spikes**.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/spikes/stress-detection/stress.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+export interface NucleusMeasurement {
+  /** Mean signal energy across the nucleus window. */
+  energy: number;
+  /** Mean fundamental frequency in Hz, or `null` where tracking failed. */
+  f0Hz: number | null;
+}
+
+export interface Nucleus {
+  /** 1-based index among the nuclei of the utterance. */
+  syllableIndex: number;
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/spikes/stress-detection/stress.ts]: Valid catch — and already addressed in `d2f4abc`. The `Nucleus` interface now carries only `phonemeIndex` (index into the spoken stream), with a doc comment explicitly stating it is deliberately the *only* index carried precisely to prevent the spoken-vs-target confusion you describe. There is no `syllableIndex` field left on the type or assigned in `extractNuclei`. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/spikes/stress-detection/stress.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+ * stressed, so this is absent evidence, not a wrong answer.
+ */
+export function mapToTargetSyllable(
+  nucleus: Nucleus,
+  alignment: readonly AlignmentEntryLike[],
+  targetPhonemes: readonly string[]
+): number | null {
+  // Entries carrying a spoken sound enumerate the spoken stream in order.
+  let spokenCursor = -1;
+  const entry = alignment.find((e) => {
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/spikes/stress-detection/stress.ts]: Valid, and adopted in `d2f4abc`. `mapToTargetSyllable` now uses the exact explicit `for` loop you suggested — `spokenCursor` is incremented by the loop body, not as a `.find()` predicate side effect — and a comment above it states the 1-to-1 spoken-stream precondition that the walk relies on. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/spikes/stress-detection/stress.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+ * mapped through the alignment (S3).
+ */
+export interface StressEvidence {
+  detected: number;
+  /** Every regionally acceptable stressed-syllable index (#2360). */
+  accepted: readonly number[];
+}
+
+/** Vowels that are reduced by definition, and so cannot carry stress (S4). */
+const REDUCED_VOWELS: ReadonlySet<string> = new Set(['ə', 'ɐ', 'ɪ̈', 'ɵ']);
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/spikes/stress-detection/stress.ts]: Valid — the `ɪ̈` (U+026A + U+0308) entry was a dead rule that could never fire against the vocabulary, and it was removed in `d2f4abc`. `REDUCED_VOWELS` is now `['ə', 'ɐ', 'ɚ', 'ᵻ']`, all four verified present in `vocab.json` by the `stress.test.ts` assertion "every reduced vowel the English cue looks for is a symbol the model can emit". Plain `ɪ` was deliberately *not* added — it is the full KIT vowel that carries stress in `bit`, so treating it as reduced would suppress stress on every syllable containing it. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/spikes/stress-detection/stress.test.ts [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+      detectStress({
+        spokenPhonemes: [tp('p', 0, 2), tp('r', 2, 4)],
+        measurements: [],
+        alignment: allCorrect(target),
+        targetPhonemes: target,
+        accepted: [2],
+        config: CONFIG,
+      })
+    ).toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/spikes/stress-detection/stress.test.ts:275]: Valid gap — the `measurements.length !== nuclei.length` guard now has dedicated coverage, added in `d2f4abc`. See `stress.test.ts` "reports nothing when the measurement count does not match the nucleus count" (two nuclei, one measurement → `null`), which is distinct from the adjacent no-nuclei test. 26/26 spike tests pass. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/spikes/stress-detection/stress.ts [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+ * executable assertions. See `stress.test.ts`.
+ *
+ * Scope guard: this detects LEXICAL STRESS ONLY — which syllable of a known
+ * word carries prominence. Fluency, rhythm and intonation remain out of scope
+ * on the map (#2331). Nothing here may grow into general prosody scoring.
+ */
+
+import nucleiData from './nuclei.json';
+
+/** The 239 nucleus-bearing tokens, derived from the model's own vocabulary. */
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply scripts/spikes/stress-detection/stress.ts]: Valid — fixed in `8fc4f39`. The comment now reads `/** The 244 nucleus-bearing tokens, derived from the model's own vocabulary. */`, matching the `stress.test.ts:62` assertion (total 244 = 239 vowel-headed + 3 syllabic consonants + 2 glide-initial) and `DECISIONS.md`. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_
